python/lsst/ts/hvac/hvac_csc.py

```diff
@@ -0,0 +1,283 @@
+
+
+class HvacCsc(salobj.ConfigurableCsc):
+    """Commandable SAL Component for the HVAC (Heating, Ventilation and Air
+    Conditioning).
+
+    Parameters
+    ----------
+    config_dir : `string`
+        The configuration directory
+    initial_state : `salobj.State`
+        The initial state of the CSC
+    simulation_mode : `int`
+        Simulation mode (1) or not (0)
+    """
+
+    def __init__(
+        self, config_dir=None, initial_state=salobj.State.STANDBY, simulation_mode=0,
+    ):
+        schema_path = pathlib.Path(__file__).resolve().parents[4].joinpath("schema", "hvac.yaml")
+        self.config = None
+        self._config_dir = config_dir
+        super().__init__(
+            name="HVAC",
+            index=0,
+            schema_path=schema_path,
+            config_dir=config_dir,
+            initial_state=initial_state,
+            simulation_mode=simulation_mode,
+        )
+        self.mqtt_client = None
+        self.telemetry_task = None
+        self.log.info("__init__")
+
+    async def connect(self):
+        """Start the HVAC MQTT client or start the mock client, if in
+        simulation mode.
+        """
+        self.log.info("Connecting")
+        self.log.info(self.config)
+        self.log.info(f"self.simulation_mode = {self.simulation_mode}")
+        if self.config is None:
+            raise RuntimeError("Not yet configured")
+        if self.connected:
+            raise RuntimeError("Already connected")
+
+        if self.simulation_mode == 1:
+            # Connect to the Simulator Client.
+            self.log.info("Connecting to SimClient.")
+            self.mqtt_client = SimClient()
+        else:
+            # Connect to the MQTT Client.
+            self.log.info("Connecting to MqttClient.")
+            self.mqtt_client = MqttClient()
+
+        self.mqtt_client.connect()
+        self.telemetry_task = asyncio.create_task(self.get_telemetry())
+
+    async def disconnect(self):
+        """Disconnect the HVAQ client, if connected.
+        """
+        self.log.info("Disconnecting")
+        if self.telemetry_task:
+            self.telemetry_task.cancel()
+        if self.connected:
+            self.mqtt_client.disconnect()
+
+    async def handle_summary_state(self):
+        """Override of the handle_summary_state function to connect or
+        disconnect to the HVAC server (or the mock server) when needed.
+        """
+        self.log.info(f"handle_summary_state {salobj.State(self.summary_state).name}")
+        if self.disabled_or_enabled:
+            if not self.connected:
+                await self.connect()
+        else:
+            await self.disconnect()
+
+    async def configure(self, config):
+        self.config = config
+
+    async def implement_simulation_mode(self, simulation_mode):
+        if simulation_mode not in (0, 1):
+            raise salobj.ExpectedError(f"Simulation_mode={simulation_mode} must be 0 or 1")
```

Comment:
I had a big discussion with Petr about this a few weeks ago and he suggested to remove it. I told him I couldn't because it was part of salobj. Thanks for putting the record straight on that.

Reply:
Actually, this makes all unit tests fail. I see in BaseCsc in salobj this method:

```
    async def implement_simulation_mode(self, simulation_mode):
        """Implement going into or out of simulation mode.

        Parameters
        ----------
        simulation_mode : `int`
            Requested simulation mode; 0 for normal operation.

        Raises
        ------
        ExpectedError
            If ``simulation_mode`` is not a supported value.
        ValueError
            If ``simulation_mode`` and ``initial_simulation_mode``
            are both nonzero. Only one can be specified.

        Notes
        -----
        Subclasses should override this method to implement simulation
        mode. The implementation should:

        * Check the value of ``simulation_mode`` and raise
          `ExpectedError` if not supported.
        * If ``simulation_mode`` is 0 then go out of simulation mode.
        * If ``simulation_mode`` is nonzero then enter the requested
          simulation mode.
        """
        if simulation_mode != 0:
            raise base.ExpectedError(
                f"This CSC does not support simulation; simulation_mode={simulation_mode} but must be 0"
            )
```
which is why I need to override that method in my CSC.

Reply:
I'm sorry that I was not clearer. You must still provide an override for implement_simulation_mode but it can be a no-op (just `pass`).

The advantage to testing for valid simulation_mode values in the CSC's constructor is that it catches the error earlier -- before the start method (a background tast) runs. It's not a big deal, just a minor enhancement.

I will file a ticket to improve support for non-zero simulation_mode in ts_salobj. There is too much boilerplate needed for every CSC with a simulation mode. I apologize for that.

Reply:
Thanks for the clarifications. I updated my code accordingly.